leaderboard/main.go

```diff
@@ -0,0 +1,122 @@
+	dsedb "dont-slack-evil/db"
+
+	"github.com/slack-go/slack"
+)
+
+// UserScore contains the positivity score of a Slack user
+type UserScore struct {
+	ID string
+	Name string
+	Good int
+	Total int
+	Score float64
+}
+
+// SendLeaderboardNotification sends the leaderboard notification
+func SendLeaderboardNotification() (int, error) {
+	notificationsSent := 0
+	teams, teamsErr := dsedb.GetTeams()
+	log.Println("Analyzing weekly messages for", len(teams), "teams")
+	if teamsErr != nil {
+		log.Println(teamsErr)
+		return 0, teamsErr
+	}
+	for _, team := range teams {
+		log.Println("Analyzing weekly messages from team", team.SlackTeamId)
+		slackBotUserApiClient := slack.New(team.SlackBotUserToken)
+		users, err := slackBotUserApiClient.GetUsers()
+		if err != nil {
+			log.Printf("Could not instantiate bot client for team %v", team.SlackTeamId)
+			continue
+		}
+		log.Println("Found", len(users), "users in team")
+		var userScores []UserScore;
+		for _, user := range users {
+			// This is the best way I found to distinguish bots from real users
+			// Note that user.IsBot doesn't work because it's false even for some bot users...
+			if (len(user.Profile.BotID) == 0 && user.Name != "slackbot") {
+				good, total := apphome.GetWeeklyStats(user.ID)
+				var score float64;
+				if (total > 0) {
+					score = float64(good) / float64(total)
+				} else {
+					score = 0
+				}
+				userScore := UserScore{
+					ID: user.ID,
+					Name: user.RealName,
+					Good: good,
+					Total: total,
+					Score: score,
+				}
+				userScores = append(userScores, userScore)
+			}
+		}
+		// Sort by positivity scores
+		sort.Slice(userScores, func(i, j int) bool {
+			return userScores[i].Score > userScores[j].Score
+		})
+		var text = "*Weekly positivity rankings:*"
+		if (len(userScores) > 0) {
+			text += fmt.Sprintf(
+				"\n\nCongratulations to <@%s> for being the most positive person this week :tada:",
+				userScores[0].ID,
+			)
+			text += "\n\nHere are the standings:"
+			text += fmt.Sprintf(
+				"\n:first_place_medal: <@%s> with a %.2f score (%d / %d)",
+				userScores[0].ID,
+				userScores[0].Score * 100,
+				userScores[0].Good,
+				userScores[0].Total,
+			)
+		}
+		if (len(userScores) > 1) {
+			text += fmt.Sprintf(
+				"\n:second_place_medal: <@%s> with a %.2f score (%d / %d)",
+				userScores[1].ID,
+				userScores[1].Score * 100,
+				userScores[1].Good,
+				userScores[1].Total,
+			)
+		}
+		if (len(userScores) > 2) {
+			text += fmt.Sprintf(
+				"\n:third_place_medal: <@%s> with a %.2f score (%d / %d)",
+				userScores[2].ID,
+				userScores[2].Score * 100,
+				userScores[2].Good,
+				userScores[2].Total,
+			)
+		}
+		params := slack.GetConversationsParameters {
+			Types: []string{"public_channel"},
+		}
+		channels, _, channelErr := slackBotUserApiClient.GetConversations(&params)
+		if (channelErr != nil) {
+			log.Println(channelErr)
+			continue
+		}
+		for _, channel := range channels {
+			if (channel.IsGeneral) {
```

Comment:
I'm not very happy with this, if you have a suggestion to choose which channel to post in, I'm all ears @Startouf @mourtisma

Reply:
Maybe the webhook channel would have been handy here... ? Otherwise, I suggest switching to the "Random" channel which would be less intrusive than general.

I believe we should make a PR asap to let users disable the weekly rankings. Via the app home maybe, or a slash command.

Reply:
Also, since you loop over all the teams, this could on the long term cause timeout issues. I have the same problem with the notification, and I will be working on it

Reply:
> Maybe the webhook channel would have been handy here... ? Otherwise, I suggest switching to the "Random" channel which would be less intrusive than general.

 I don't think #random would be best because there will always been one general channel but not necessarily a random channel.

I agree with using a webhook channel, I will do that

> I believe we should make a PR asap to let users disable the weekly rankings. Via the app home maybe, or a slash command.

Created https://github.com/gjgd/dont-slack-evil/issues/53 to track that

> Also, since you loop over all the teams, this could on the long term cause timeout issues. I have the same problem with the notification, and I will be working on it

I haven't encountered this problem, but it is possible to increase the default serverless timeout of 6s (see https://serverless.com/framework/docs/providers/aws/guide/functions/)

Reply:
Implemented using webhook